bpfd/src/bpf.rs

```diff
@@ -991,17 +1007,51 @@ impl BpfManager {
         Ok(())
     }
 
-    fn rebuild_map_entry(&mut self, id: u32, map_owner_id: Option<u32>) {
+    fn rebuild_map_entry(&mut self, id: u32, program: &mut Program) {
+        let map_owner_id = match program.data() {
+            Ok(data) => data.map_owner_id(),
+            Err(_) => {
+                warn!("unable to retrieve data for {id} retrieving map_owner_id on rebuild");
+                return;
+            }
+        };
         let index = match map_owner_id {
             Some(i) => i,
             None => id,
         };
 
         if let Some(map) = self.maps.get_mut(&index) {
             map.used_by.push(id);
+
+            // This program has not been inserted yet, so update it with the
+            // updated map_used_by.
+            if let Ok(data) = program.data_mut() {
+                data.set_maps_used_by(Some(map.used_by.clone()));
+            } else {
+                warn!("unable to retrieve data for {id} during rebuild of maps");
+            }
+
+            // Update all the other programs using the same map with the updated map_used_by.
+            for used_by_id in map.used_by.iter() {
+                // program may not exist yet on rebuild, so ignore if not there
+                if let Some(prog) = self.programs.get_mut(used_by_id) {
+                    if let Ok(data) = prog.data_mut() {
+                        data.set_maps_used_by(Some(map.used_by.clone()));
+                    } else {
+                        warn!("unable to retrieve data for {used_by_id} when setting map_used_by on rebuild");
+                    }
+                }
+            }
         } else {
             let map = BpfMap { used_by: vec![id] };
             self.maps.insert(index, map);
+
+            // Update this programs with the updated map_used_by.
```

Comment:
nit: I don't think you need this comment.
```suggestion
```

Reply:
Done